opm/ml/ml_model.hpp

```diff
@@ -0,0 +1,437 @@
+/*
+  Copyright (c) 2016 Robert W. Rose
+  Copyright (c) 2018 Paul Maevskikh
+  Copyright (c) 2024 NORCE
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef ML_MODEL_H_
+#define ML_MODEL_H_
+
+#include <fmt/format.h>
+
+#include <algorithm>
+#include <chrono>
+#include <cmath>
+#include <cstdio>
+#include <numeric>
+#include <opm/common/ErrorMacros.hpp>
+#include <opm/material/densead/Math.hpp>
+#include <string>
+#include <vector>
+
+namespace Opm {
+
+// NN layer
+// ---------------------
+/** \class Tensor class
+ * Implements mathematical tensor (Max 4d)
+ */
+template <class T> class Tensor {
+ public:
+  Tensor() {}
+
+  explicit Tensor(int i) { resizeI<int>({i}); }
+
+  Tensor(int i, int j) { resizeI<int>({i, j}); }
+
+  Tensor(int i, int j, int k) { resizeI<int>({i, j, k}); }
+
+  Tensor(int i, int j, int k, int l) { resizeI<int>({i, j, k, l}); }
+
+  template <typename Type> void resizeI(std::vector<Type> c) {
+    if (c.size() == 1) dims_ = {(int)c[0]};
+    if (c.size() == 2) dims_ = {(int)c[0], (int)c[1]};
+    if (c.size() == 3) dims_ = {(int)c[0], (int)c[1], (int)c[2]};
+    if (c.size() == 4) dims_ = {(int)c[0], (int)c[1], (int)c[2], (int)c[3]};
+
+    data_.resize(
+        std::accumulate(begin(c), end(c), 1.0, std::multiplies<Type>()));
+  }
+
+  inline void flatten() {
```

Comment:
Remove `inline` in front of functions that are defined inline in the class. It is unnecessary.

This was commented on already!